ffi/pic.lua

```diff
@@ -272,6 +272,9 @@ function Pic.openPNGDocument(filename)
     elseif re.ncomp == 3 then bbtype = BB.TYPE_BBRGB24
     elseif re.ncomp == 4 then bbtype = BB.TYPE_BBRGB32
     else
+        if re.data then
```

Comment:
Probably a good idea, as unlikely as this should be ;).

Reply:
Well, valgring kept reporting this:
```
1,440,000 bytes in 1 blocks are possibly lost in loss record 3,972 of 3,980
   at 0x483577F: malloc (in .../vgpreload_memcheck-amd64-linux.so)
   by 0x10FB445A: lodepng_malloc (lodepng.c:78)
   by 0x10FBCA6F: lodepng_decode (lodepng.c:4899)
   by 0x12BCF5: lj_vm_ffi_call (buildvm_x86.dasc:2575)
   by 0x149F19: lj_ccall_func (lj_ccall.c:1161)
   by 0x127E8D: lj_cf_ffi_meta___call (lib_ffi.c:230)
   by 0x129D56: lj_BC_FUNCC (buildvm_x86.dasc:809)
   by 0x11A03F: lua_call (lj_api.c:1111)
```
Looking at (your) ffi code, it seems we do things right by passing that data to BlitBuffer and using setAllocated(1) - so it should properly be free. Dunno why valgrind kept telling that one - and mostly only that one that feels serious (1.4 Mb!)

Reply:
Any idea on which pixmap that was happening?

Nothing should ever return anything than [1..4) in ncomp. If it does, that could arguably mean there's a real bug somewhere ;).

Reply:
Nope. Can't have valgrind report that again now... although I think I got them even after that little fix.

Reply:
Mhhh, got it again. It may be logged at the end of exiting koreader/valgring, so dunno where it comes from. I did "Refresh cached book information" on a EPUB with a cover, although this should use MuPDF for that.
I added that:
```diff
--- a/ffi/pic.lua
+++ b/ffi/pic.lua
@@ -262,18 +262,20 @@ function Pic.openPNGDocument(filename)
         -- NOTE: LodePNG will NOT do RGB -> Grayscale conversions for us, for design reasons (multiple ways to do it, lossy).
         --       So we can only *ask* to keep grayscale PNGs as-is, but we might actually be getting fed back a RGB24 one ;).
         req_n = 1
     end

+    print("pngopen: "..filename)
     local ok, re = Png.decodeFromFile(filename, req_n)
     if not ok then error(re) end

     if re.ncomp == 1 then bbtype = BB.TYPE_BB8
     elseif re.ncomp == 2 then bbtype = BB.TYPE_BB8A
     elseif re.ncomp == 3 then bbtype = BB.TYPE_BBRGB24
     elseif re.ncomp == 4 then bbtype = BB.TYPE_BBRGB32
     else
+        print("pngfailfree: "..filename)
         if re.data then
             ffi.C.free(re.data)
         end
         error("unsupported number of color components")
     end
```
which looks to me the only place that could call `lodepng_decode`, and I didn't see any of my print printed once !
It's from ffi, so not from thirdparty/ stuff that could use lodepng themselves... Quite puzzled.